add dakota instructions (#102)
MaMMoS-project/mammos: examples/dakota.rst

Using external optimizers (e.g. Dakota)
=======================================

This page documents how an external optimizer such as Dakota can be implemented
together with the mammos package in the :doc:`sensor`.

We assume the user has already installed a version of Dakota compatible with the
MaMMoS software suite, i.e. compiled with Python 3.11. The compilation must have
included all the necessary Python flags (for further information, see `here <https://snl-dakota.github.io/docs/6.24.0/users/usingdakota/advanced/advancedsimulationcodeinterfaces.html#python>`__).

> For help on compiling Dakota, follow the guide in https://github.com/MaMMoS-project/dakota-sensor-optimization.

Reviewer: I would maybe drop the word `help` and instead say something like "an example/recipe for compiling Dakota <insert version> on Linux can be found in ..."

> 

Reviewer: At this point it would be good to briefly repeat the overall aim of this tutorial, i.e. shape optimization of a sensor under certain constraints. We could probably copy-paste this from the notebook.

Author: I have added a small description in the first paragraph without pasting it from the previous notebook, please check if it looks good.

To use Dakota's Python interface, first we need to define a Python function
accepting inputs and creating output in the format compatible with Dakota.
In particular, this function must accept a dictionary input where the item ``cv``
holds the continuous variables. The output is expected to be a dictionary with

> the items ``fns`` (list of objective functions), ``fnGrads`` (the list of gradients),

Reviewer: Is it really a list of functions or list of results of function evaluations at `cv`?

and ``fnHessians`` (list of Hessians), if needed. See :ref:`sensor-py` for an example
reproducing the optimization problem in :doc:`sensor`.

> 

Reviewer: We should add another paragraph that explains `sensor.py` in detail. In particular: we cannot expect any familiarity with Ubermag.

Author: I have added a brief summary of the function where we use Ubermag without describing the Ubermag objects.

.. code-block:: python
   :caption: Content of ``sensor.py``
   :name: sensor-py

   import dakota.interfacing as di
   import discretisedfield as df
   import mammos_analysis as ma
   import mammos_entity as me
   import mammos_units as u
   import micromagneticmodel as mm
   import oommfc as mc

   def optimize(params):
       # read parameters
       sx, sy = params["cv"]

       # geometry and mesh
       L = 100e-9  # nm
       t = 5e-9  # nm
       region = df.Region(p1=(-L / 2, -L / 2, -t / 2), p2=(L / 2, L / 2, t / 2))
       mesh = df.Mesh(region=region, n=(40, 40, 1))

       # intrinsic properties
       A = me.A(6.363746956380124e-12, unit="J/m")

>        Ms = me.Ms(831.7446855583435, "kA/m")

Reviewer: That's too many digits in my opinion.

       # hysteresis loop settings and driver initialization
       hd = mc.HysteresisDriver()
       Hmin = (0, 0, 0)
       Hmax = ((0.1, 500, 0) * u.mT).to("A/m", equivalencies=u.magnetic_flux_field())
       n = 101

       # system
       system = mm.System(name="sensor")
       system.energy = mm.Exchange(A=A.value) + mm.Demag() + mm.Zeeman(H=(0, 0, 0))

       # define diamond mask
       def in_diamond(position):
           x, y, _ = position
           if abs(x) / sx + abs(y) / sy <= 1:
               return Ms.q.to("A/m").value
           else:
               return 0

       # magnetization
       system.m = df.Field(mesh, nvdim=3, value=(1, 0, 0), norm=lambda p: in_diamond(p), valid="norm")

       # run hysteresis loop and return linearization value
       hd.drive(system, Hsteps=[[Hmin, tuple(Hmax.value), n]], verbose=0)  # Field sweep

>        H_y = me.H(system.table.data["By_hysteresis"].values * u.Unit(system.table.units["By_hysteresis"]).to("A/m", equivalencies=u.magnetic_flux_field()))

Reviewer: This line is too long.

       M_y = system.table.data["my"].values * Ms.q
       margin = 0.05 * Ms.q.to("A/m")
       res = ma.hysteresis.find_linear_segment(H_y, M_y, margin=margin, min_points=2)
       return {"fns": [res.Hmax.value]}

A possible Dakota input file is proposed in :ref:`sensor-in`.
We have used the `efficient global optimization method <https://snl-dakota.github.io/docs/6.24.0/users/usingdakota/theory/surrogatebasedglobaloptimization.html#efficient-global-optimization>`__,

> which can be an overkill for this problem, with 10 initial random samples

Reviewer: Not sure if we want to keep this statement.

   :name: sensor-out

   Dakota version 6.24 released May 15 2026.
   Repository revision 1ab1c9924 (2026-05-11) built Jun  8 2026 15:35:45.
   Running serial Dakota executable in serial mode.
   Start time: Thu Jun 18 10:14:34 2026

   ...

   Stopping criteria not met: distConvergenceCntr (0) < distConvergenceLimit (1)
   Stopping criteria not met: eifConvergenceCntr (1) < eifConvergenceLimit (2)
   Stopping criteria met:     globalIterCount (20) >= maxIterations (20)
   <<<<< Function evaluation summary: 30 total (30 new, 0 duplicate)
   <<<<< Best parameters          =
                         6.9999978988e-08 x1
                         3.0000210121e-09 x2
   <<<<< Best objective function  =
                         3.9788735730e+05
   <<<<< Best evaluation ID: 12

   <<<<< Iterator efficient_global completed.
   <<<<< Environment execution completed.
   DAKOTA execution time in seconds:
     Total CPU        =    19.4231 [parent =    19.4231, child =      3e-06]
     Total wall clock =    73.4983

It is possible to store the essential output information by calling, for example
:code:`dakota -i sensor.in -o sensor.out` (and you might want to redirect stdout by

> adding ``> sensor.stdout`` at the end of the previous command).

Reviewer: I think we should mention at least the `-o sensor.out` in the paragraph above, before showing the output; maybe even both. Someone following this tutorial would already have run Dakota only to learn at this point that basically no (?) output has been saved.

Author: Good idea, I have done as you suggested, please take a look.

> In particular, the output file ``sensor.out`` contains the input file at the top

Reviewer: Maybe it would be more useful to show `sensor.out` instead of the stdout of the dakota call.

Author: This is `sensor.out`. Is it unclear from the text?

Reviewer: It wasn't when I first read it but it is clear in the latest version.

> of the tile and the optimization results at the end. This includes information such

Reviewer: ```suggestion
of the file and the optimization results at the end. This includes information such
```

maximum), the ``Best parameters`` (the maximizing couple), and the execution time.
For more information about the output, see the `Dakota Output documentation
<https://snl-dakota.github.io/docs/6.24.0/users/usingdakota/output.html>`__.

In :ref:`out-py` we include a Python script to load the optimization information from
the Dakota output into a mammos-yaml file.

.. code-block::
   :caption: Script to store Dakota output into a mammos-yaml file
   :name: out-py

   import mammos_entity as me

   with open("sensor.out") as f:
       lines = f.readlines()
   for i, line in enumerate(lines):
       if "<<<<< Best parameters" in line:
           x1 = float(lines[i+1].strip().split()[0])
           x2 = float(lines[i+2].strip().split()[0])
       if "<<<<< Best objective function" in line:
           opt = float(lines[i+1].strip())

   me.EntityCollection(
       description="Optimization result in order to maximize linear response of rhomboydal sensor.",
       x1=me.Entity("Length", x1, description="optimal sensor x-semiaxis"),
       x2=me.Entity("Length", x2, description="optimal sensor y-semiaxis"),

>        linear_response=opt,

Reviewer: What's the meaning of `linear_response` resp. `opt`? That's not at all clear to me.

Author: I have tried compressing this information in the collection description. Hopefully the description plus the problem explanation at the top of the file help.

   ).to_yaml("sensor.yaml")

In :ref:`sensor-yaml` we can see the content of the generated file.

> .. code-block::

Reviewer: ```suggestion
.. code-block:: yaml
```

   :caption: Content of ``sensor.yaml``
   :name: sensor-yaml

   # mammos yaml v2
   metadata: null
   description: Optimization result in order to maximize linear response of rhomboydal
     sensor.
   data:
     x1:
       ontology_label: Length
       description: optimal sensor x-semiaxis
       ontology_iri: https://w3id.org/emmo#EMMO_cd2cd0de_e0cc_4ef1_b27e_2e88db027bac
       unit: m
       value: 6.9999978988e-08
     x2:
       ontology_label: Length
       description: optimal sensor y-semiaxis
       ontology_iri: https://w3id.org/emmo#EMMO_cd2cd0de_e0cc_4ef1_b27e_2e88db027bac
       unit: m
       value: 3.0000210121e-09
     linear_response:
       value: 397887.3573

> Observe that each line in the ``Best parameters`` section in the Dakota

Reviewer: ```suggestion
Note that each line in the ``Best parameters`` section in the Dakota
```

Using external optimizers (e.g. Dakota)
=======================================

This page documents how an external optimizer such as Dakota can be implemented
together with the mammos package in the :doc:`sensor`. The optimization problem
consists into finding the optimal shape of a sensor in order to maximize the
linear section of the hysteresis loop. As we are assuming the sensor to be a

> rhombus, we will try to optimize the two axes here called ``sx`` and ``sy``.

Reviewer: I would align this more with the sensor shape optimization notebook

```suggestion
Sensor shape optimization with Dakota
=====================================

This example documents how an external optimizer such as Dakota can be used
together with the mammos framework. We consider the same optimization problem
as in the :doc:`sensor`: finding the optimal shape of a rhombohedral sensor to maximize the
linear section of the hysteresis loop. We well optimize the lengths of the two axes of
the rhombus, here called ``sx`` and ``sy``.
```

We assume the user has already installed a version of Dakota compatible with the
MaMMoS software suite, i.e. compiled with Python 3.11. The compilation must have
included all the necessary Python flags (for further information, see `here <https://snl-dakota.github.io/docs/6.24.0/users/usingdakota/advanced/advancedsimulationcodeinterfaces.html#python>`__).

A recipe for compiling Dakota on Linux can be found in https://github.com/MaMMoS-project/dakota-sensor-optimization.

To use Dakota's Python interface, first we need to define a Python function
accepting inputs and creating output in the format compatible with Dakota.
In particular, this function must accept a dictionary input where the item ``cv``
holds the continuous variables. The output is expected to be a dictionary with
the items ``fns`` (list of objective function values evaluated at the
continuous variables), ``fnGrads`` (the list of gradient values), and
``fnHessians`` (list of Hessian values), if needed. See :ref:`sensor-py` for
an example reproducing the optimization problem in :doc:`sensor`.

The function ``optimize`` interacts directly with the Dakota interface, while

> the function ``objective`` defines and simulation the micromagnetic system. In

Reviewer: ```suggestion
the function ``objective`` defines the micromagnetic simulation. In
```

particular, it builds the geometry, defines the intrinsic properties and runs

> an hysteresis loop. Then it analyses the loop and extract the length of the

Reviewer: ```suggestion
a hysteresis loop. Then it analyses the loop and extracts the length of the
```

> linear sector, returning it as the value of the objective function.

Reviewer: ```suggestion
linear segment, returning it as the value of the objective function.
```

.. code-block:: python
   :caption: Content of ``sensor.py``
   :name: sensor-py

   import dakota.interfacing as di
   import discretisedfield as df
   import mammos_analysis as ma
   import mammos_entity as me
   import mammos_units as u
   import micromagneticmodel as mm
   import oommfc as mc

   def optimize(params):
       # read parameters
       sx, sy = params["cv"]
       # evaluate objective function
       res = objective(sx, sy)
       # return value

>        return {"fns": [res.Hmax.value]}

Reviewer: The comments as they are, are of limited use. I also don't like the function name `objective` as this is too generic in my opinion

```suggestion
       # extract rhombus axes from parameters dictionary passed by Dakota
       sx, sy = params["cv"]
       # evaluate objective function
       properties = compute_linear_segment_properties(sx, sy)
       # return the maximum field strength in the linear segment
       return {"fns": [properties.Hmax.value]}
```

>    def objective(sx, sy):

Reviewer: ```suggestion
   def compute_linear_segment_properties(sx, sy):
   """Compute linear segment for a rhombohedral sensor.
   
   Parameters ``sx`` and ``sy`` are the lengths of the two
   axes in metre.
   """
   # This function uses Ubermag to perform the micromagnetic
   # simulations and mammos_analysis to extract properties
   # of the linear segment of the hysteresis loop.
```

       # geometry and mesh
       L = 100e-9  # nm
       t = 5e-9  # nm
       region = df.Region(p1=(-L / 2, -L / 2, -t / 2), p2=(L / 2, L / 2, t / 2))
       mesh = df.Mesh(region=region, n=(40, 40, 1))

       # intrinsic properties
       A = me.A(6e-12, unit="J/m")
       Ms = me.Ms(832, "kA/m")

       # hysteresis loop settings and driver initialization
       hd = mc.HysteresisDriver()
       Hmin = (0, 0, 0)
       Hmax = ((0.1, 500, 0) * u.mT).to("A/m", equivalencies=u.magnetic_flux_field())
       n = 101

       # system
       system = mm.System(name="sensor")

>        system.energy = mm.Exchange(A=A.value) + mm.Demag() + mm.Zeeman(H=(0, 0, 0))

Reviewer: ```suggestion
       system.energy = mm.Exchange(A=A.value) + mm.Demag()
```
The Zeeman energy term is unused and seems to be some leftover.

       # define diamond mask
       def in_diamond(position):
           x, y, _ = position
           if abs(x) / sx + abs(y) / sy <= 1:
               return Ms.q.to("A/m").value
           else:
               return 0

       # magnetization
       system.m = df.Field(mesh, nvdim=3, value=(1, 0, 0), norm=lambda p: in_diamond(p), valid="norm")

       # run hysteresis loop and return linearization value
       hd.drive(system, Hsteps=[[Hmin, tuple(Hmax.value), n]], verbose=0)  # Field sweep
       B_y = system.table.data["By_hysteresis"].values * u.Unit(system.table.units["By_hysteresis"])
       H_y = me.H(B_y.to("A/m", equivalencies=u.magnetic_flux_field()))
       M_y = system.table.data["my"].values * Ms.q
       margin = 0.05 * Ms.q.to("A/m")

>        return ma.hysteresis.find_linear_segment(H_y, M_y, margin=margin, min_points=2)

Reviewer: ```suggestion
       # run hysteresis loop simulation
       hd.drive(system, Hsteps=[[Hmin, tuple(Hmax.value), n]], verbose=0)
       
       # read simulation results and extract properties of linear segment
       B_y = system.table.data["By_hysteresis"].values * u.Unit(system.table.units["By_hysteresis"])
       H_y = me.H(B_y.to("A/m", equivalencies=u.magnetic_flux_field()))
       M_y = system.table.data["my"].values * Ms.q
       margin = 0.05 * Ms.q.to("A/m")
       return ma.hysteresis.find_linear_segment(H_y, M_y, margin=margin, min_points=2)
```

> A possible Dakota input file is proposed in :ref:`sensor-in`.

Reviewer: ```suggestion
A possible Dakota input file is shown in :ref:`sensor-in`.
```

In the ``variables`` section we define the number of continuous variables, the
initial points and the bounds. The ``interface`` section specifies the
Python function ``sensor.optimize`` (i.e. the function defined in
:ref:`sensor-py`) to be optimized. The last section, ``responses``, includes
information about the number of objective functions and whether we are calculating
the gradients and/or the Hessian. Extensive documentation about all keywords is
available in the `Dakota Keyword Reference <https://snl-dakota.github.io/docs/6.24.0/users/usingdakota/reference.html#>`__.

.. code-block::
   :caption: Content of ``sensor.in``
   :name: sensor-in

   method
     efficient_global
       initial_samples    10
       max_iterations     20

   variables
     continuous_design = 2
       cdv_initial_point  19e-9  39e-9
       cdv_lower_bounds    3e-9   3e-9
       cdv_upper_bounds    7e-8   7e-8
       cdv_descriptor       'sx'  'sy'

   interface
     analysis_driver = 'sensor.optimize'
       python

   responses
     objective_functions = 1
       sense = "maximization"
     no_gradients
     no_hessians

> Dakota can then be called by the command :code:`dakota -i sensor.in -o sensor.out`.

Reviewer: ```suggestion
Dakota can then be called on the command line using :code:`dakota -i sensor.in -o sensor.out`.
```

particular, it builds the geometry, defines the intrinsic properties and runs
a hysteresis loop. Then it analyses the loop and extracts the length of the
linear segment, returning it as the value of the objective function.

.. code-block:: python
   :caption: Content of ``sensor.py``
   :name: sensor-py

   import dakota.interfacing as di
   import discretisedfield as df
   import mammos_analysis as ma
   import mammos_entity as me
   import mammos_units as u
   import micromagneticmodel as mm
   import oommfc as mc

   def optimize(params):
       # extract rhombus axes from parameters dictionary passed by Dakota
       sx, sy = params["cv"]
       # evaluate objective function
       properties = compute_linear_segment_properties(sx, sy)
       # return the maximum field strength in the linear segment
       return {"fns": [properties.Hmax.value]}

   def compute_linear_segment_properties(sx, sy):
   """Compute linear segment for a rhombohedral sensor.

   Parameters ``sx`` and ``sy`` are the lengths of the two
   axes in metre.
   """
   # This function uses Ubermag to perform the micromagnetic
   # simulations and mammos_analysis to extract properties

>    # of the linear segment of the hysteresis loop.

Reviewer: Wrong indentation.

   <<<<< Best evaluation ID: 11

   <<<<< Iterator efficient_global completed.
   <<<<< Environment execution completed.
   DAKOTA execution time in seconds:
     Total CPU        =    19.3955 [parent =    19.3955, child =      6e-06]
     Total wall clock =    87.6421

In :ref:`out-py` we include a Python script to load the optimization information from
the Dakota output into a mammos-yaml file.

.. code-block::
   :caption: Script to store Dakota output into a mammos-yaml file
   :name: out-py

   import mammos_entity as me

   with open("sensor.out") as f:
       lines = f.readlines()
   for i, line in enumerate(lines):
       if "<<<<< Best parameters" in line:
           sx = float(lines[i+1].strip().split()[0])
           sy = float(lines[i+2].strip().split()[0])
       if "<<<<< Best objective function" in line:
           linear_response = float(lines[i+1].strip())

   me.EntityCollection(
       description="Optimized rhomboydal sensor maximizing the linear section of the hysteresis loop.",
       sx=me.Entity("Length", sx, description="optimal sensor x-semiaxis"),
       sy=me.Entity("Length", sy, description="optimal sensor y-semiaxis"),

>        linear_response=linear_response,

Reviewer: `linear_response` is still not clear.

```suggestion
       linear_response=me.Entity("ExternalMagneticField", linear_response, description="maximum field strength in the linear segment"),
```

The suggestion is based on the property LinearSegmentProperties.Hmax (https://mammos-project.github.io/mammos/api/mammos_analysis.hysteresis.html#mammos_analysis.hysteresis.LinearSegmentProperties), which I believe `linear_response` actually is?

   ).to_yaml("sensor.yaml")

In :ref:`sensor-yaml` we can see the content of the generated file.

.. code-block:: yaml
   :caption: Content of ``sensor.yaml``
   :name: sensor-yaml

   # mammos yaml v2
   metadata: null
   description: Optimization result in order to maximize linear response of rhomboydal
     sensor.
   data:
     sx:
       ontology_label: Length
       description: optimal sensor x-semiaxis
       ontology_iri: https://w3id.org/emmo#EMMO_cd2cd0de_e0cc_4ef1_b27e_2e88db027bac
       unit: m
       value: 6.9999978988e-08
     sy:
       ontology_label: Length
       description: optimal sensor y-semiaxis
       ontology_iri: https://w3id.org/emmo#EMMO_cd2cd0de_e0cc_4ef1_b27e_2e88db027bac
       unit: m
       value: 3.0000210121e-09
     linear_response:

>        value: 397887.3573

Reviewer: Please update this to reflect the change in the python script above. It should now be an entity.
